smac/optimizer/ei_optimization.py

```diff
@@ -443,25 +471,35 @@ class ChallengerList(object):
         ConfigurationSpace from which to sample new random configurations.
     """
 
-    def __init__(self, challengers, configuration_space):
+    def __init__(self, challengers, configuration_space, random_configuration_chooser=ChooserNoCoolDown(1)):
         self.challengers = challengers
         self.configuration_space = configuration_space
         self._index = 0
-        self._next_is_random = False
+        self._last_was_random = True
+        self.random_configuration_chooser = random_configuration_chooser
 
     def __iter__(self):
         return self
 
-    def __next__(self):
-        if self._index == len(self.challengers) and not self._next_is_random:
-            raise StopIteration
-        elif self._next_is_random:
-            self._next_is_random = False
+    def _sample(self, rand):
+        self._last_was_random = rand
+        if rand:
             config = self.configuration_space.sample_configuration()
             config.origin = 'Random Search'
-            return config
         else:
-            self._next_is_random = True
             config = self.challengers[self._index]
             self._index += 1
-            return config
+        return config
+
+    def __next__(self):
+
+        if self._index == len(self.challengers):
+            if self._last_was_random:
+                raise StopIteration
+            else:
+                # finish with random configuration
+                return self._sample(rand=True)
+        elif not self._last_was_random and self.random_configuration_chooser.check(self._index):
```

Comment:
We could also directly ask the chooser whether self._index should be a random configuration or not, e.g. by setting modulus to 2? Or am I missing something?

This would make this code much cleaner.

Reply:
I wanted to stay as close as possible to the original implementation so I did not change this detail. Note the change you propose does not preserve the last configuration to be a random one.
Despite of that I refactored the code according to your suggestion.

Reply:
@mlindauer: See ^Is this behaviour crucial?